### PR TITLE
Refactored SnackbarMessageQueue to use Dispatcher

### PR DESCRIPTION
**Pull request by greuelpirat**
I reworked the `SnackbarMessageQueue` to no longer start its own long running task. Instead it uses the `Dispatcher` to _pull_ for next item.

That makes it necessary that a new instance of `SnackbarMessageQueue` must be created from within a `Dispatcher`-thread. But this should be the standard case any way. But it is a new limitation.

This relates to #1878. Although I could not reproduce it. But I personally did not like it anyway, that the `SnackbarMessageQueue` requires its own thread.

### CODE REVIEW

**Comment by bombipappoo** [comment]
I think that `_pairedSnackbarsLock` is no longer needed since `_pairedSnackbars` is only be accessed on the dispatcher thread.

File: MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs [hunk 1]
```diff
@@ -152,11 +136,12 @@ public SnackbarMessageQueue()
         public SnackbarMessageQueue(TimeSpan messageDuration)
         {
             _messageDuration = messageDuration;
-            Task.Factory.StartNew(PumpAsync, TaskCreationOptions.LongRunning);
+            _dispatcher = Dispatcher.FromThread(Thread.CurrentThread)
+                          ?? throw new InvalidOperationException("SnackbarMessageQueue must be created in a dispatcher thread");
```
Comment (w-syss) [inline MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs:140]: Why are we not using Application.Current.Dispatcher?

File: MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs [hunk 2]
```diff
@@ -68,10 +57,10 @@ private async void UiElementOnMouseLeave(object sender, MouseEventArgs mouseEven
                     }
                     catch (ObjectDisposedException)
                     {
-                        /* we are we suppresing this? 
+                        /* we are we suppressing this? 
                          * as we have switched out wait onto another thread, so we don't block the UI thread, the
                          * _cleanUp/Dispose() action might also happen, and the _disposedWaitHandle might get disposed
-                         * just before we WaitOne. We wond add a lock in the _cleanUp because it might block for 2 seconds.
+                         * just before we WaitOne. We want add a lock in the _cleanUp because it might block for 2 seconds.
```
Comment (w-syss) [inline MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs:63]: I think this was supposed to say won't :D

File: MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs [hunk 3]
```diff
@@ -80,66 +69,61 @@ private async void UiElementOnMouseLeave(object sender, MouseEventArgs mouseEven
...
                 //this keeps the event waiting simpler, rather that actually watching play -> pause -> play -> pause etc
                 var granularity = TimeSpan.FromMilliseconds(200);
 
                 Task.Factory.StartNew(() =>
                 {
                     //keep upping the completion time in case it's paused...
-                    while (DateTime.Now < _completionTime && !ceaseWaitHandle.WaitOne(granularity)
-                        && !signalWhenDurationPassedWaitHandle.WaitOne(TimeSpan.Zero))
+                    while (DateTime.Now < completionTime && !ceaseWaitHandle.WaitOne(granularity)
+                                                          && !signalWhenDurationPassedWaitHandle.WaitOne(TimeSpan.Zero))
```
Comment (w-syss) [inline MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs:115]: Maybe extract that condition with a fitting name?

File: MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs [hunk 4]
```diff
@@ -155,8 +139,8 @@ public SnackbarMessageQueue(TimeSpan messageDuration)
             Task.Factory.StartNew(PumpAsync, TaskCreationOptions.LongRunning);
         }
 
-        //oh if only I had Disposable.Create in this lib :)  tempted to copy it in like dragabalz, 
-        //but this is an internal method so no one will know my direty Action disposer...
+        //oh if only I had Disposable.Create in this lib :)  tempted to copy it in like dragablz, 
+        //but this is an internal method so no one will know my direct Action disposer...
```
Comment (bombipappoo) [inline MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs:144]: I think this was "dirty".

File: MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs (hunk 1 above)
Comment (greuelpirat) [reply MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs:140]: `Application.Current.Dispatcher` is not clear in applications with multiple dispatchers (I have such in my application).

I took this from the `DispatcherObject` where it gets its dispatcher with `Dispatcher.CurrentDispatcher`. But this returns a new Dispatcher of the current does not have one. See [here](https://referencesource.microsoft.com/#windowsbase/Base/System/Windows/Threading/Dispatcher.cs).

If you like I can change it to `Dispatcher.CurrentDispatcher`

File: MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs (hunk 2 above)
Comment (greuelpirat) [reply MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs:63]: sure, you are right. I did not really read the comments. just picked the first spell-check suggestion :-)

File: MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs (hunk 4 above)
Comment (greuelpirat) [reply MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs:144]: sure, I will change that

File: MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs (hunk 3 above)
Comment (greuelpirat) [reply MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs]: I updated this method

**Comment by greuelpirat** [comment]
Just for better understanding. Now every `SnackbarMessageQueue.Enqueue` call invokes `ShowNextAsync` on the dispatcher. First step there is to await the semaphore signal:

https://github.com/greuelpirat/MaterialDesignInXamlToolkit/blob/76212a5a2af3314f20898f9cd771f907cde2e194/MaterialDesignThemes.Wpf/SnackbarMessageQueue.cs#L256

The semaphore is set to only have one signal at a time. When the message is completed the semaphore releases this signal for the next message. So when 10 messages are queued there a 10 tasks where each task picks the next message (not the one it triggered) to show.

**Comment by bombipappoo** [comment]
If we have multiple dispatchers e.g., dispatcherA and dispatcherB, on the context of dispatcherB, the message queue created on dispatcherA can be assigned to the snackbar created on dispatcherB.
This is incorrect, so while setting `Snackbar.MessageQueue` `InvalidOperationException` should be thrown instead.

**Pull request merged**